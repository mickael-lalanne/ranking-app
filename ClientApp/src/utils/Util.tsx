import React from "react";

// All items that are not yet saved in the database starts with the following identifier
const TMP_ID_IDENTIFIER = '99999';

/**
 * Generate a temporary for items that are not yet saved in database
 * @returns {string} a random temporary id
 */
export const generateRandomId = (): string => {
    const randomNumber: number = Math.floor(Math.random() * 1000000000);
    return TMP_ID_IDENTIFIER + randomNumber;
};

/**
 * Check if an id is temporary
 * @param {string} id the id to check
 * @returns {boolean} true if the id is temporary
 */
export const isTemporaryId = (id: string): boolean => {
    return id.startsWith(TMP_ID_IDENTIFIER);
}

export interface ResizedImage {
    source: string; // the data URL of the image
    name: string;
}

/**
 * Resize an image while keeping its ratio
 * This function is called in the "onChange" event of input file
 * So we use the FileReader api to get the base image dimensions
 * @param {File} imageFile the image file to resize
 * @returns {Promise<ResizedImage>} an object containing the resized image as data URL and its name
 */
export const resizeImage = (imageFile: File): Promise<ResizedImage> => {
    const MAX_SIZE: number = 150;
    const reader = new FileReader();

    return new Promise ((resolve) => {
        reader.onload = () => {
            const tmpImg: HTMLImageElement = new Image();
    
            tmpImg.onload = () => {
                const resizeRatio = Math.min(MAX_SIZE / tmpImg.width, MAX_SIZE / tmpImg.height);

                // Initialize the canvas
                const canvas = document.createElement("canvas");
                const ctx = canvas.getContext('2d');
    
                // Set width and height
                canvas.width = tmpImg.width * resizeRatio;
                canvas.height = tmpImg.height * resizeRatio;
    
                // Draw image and export to a data-uri
                ctx!.drawImage(tmpImg, 0, 0, canvas.width, canvas.height);
    
                const dataURI = canvas.toDataURL();
                resolve({ source: dataURI, name: imageFile.name });
            };

            tmpImg.src = URL.createObjectURL(imageFile);
        }
        reader.readAsDataURL(imageFile);
    });

};

/**
 * @param array an array containing a createdAt field
 * @returns {Partial<{ createdAt: string }>[]} the same array but sorted by creation date
 */
export const sortByCreationDate = (array: Partial<{ createdAt: string }>[]) => {
    // Use slice() to avoid reference errors
    return array.slice().sort((a, b) => 
        new Date(a.createdAt!).getTime() - new Date(b.createdAt!).getTime()
    );
};

/**
 * Called in the template and tierlist editors
 * Returns a value for the MUI Tooltip component
 * @param {string?} content the text that should be displayed in the tooltip 
 * @returns {'' | React.JSX.Element | undefined} a span including the text if defined, otherwise undefined
 */
export const getTooltipTitleForSaveButtons = (
    content?: string
): '' | React.JSX.Element | undefined => {
    return content && <span style={{ fontSize: '15px'}}>{content}</span>;
}
