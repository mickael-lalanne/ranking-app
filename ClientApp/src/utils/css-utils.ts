export const ELEMENT_SIZE: string = '75px';
