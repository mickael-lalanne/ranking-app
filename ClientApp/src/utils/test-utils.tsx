import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
