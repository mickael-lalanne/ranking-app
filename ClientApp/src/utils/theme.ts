import { Theme, createTheme, alpha, ThemeOptions } from "@mui/material";

interface RankingTheme {
    primary: string;
    primaryLight: string;
    light: string;
    dark: string;
    info: string;
};
const PRIMARY_COLOR = '#BB2525';

declare module '@mui/material/styles' {
    interface Theme {
        defaultRankingTheme: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        defaultRankingTheme?: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
}

export const THEMES_COLORS: ThemeOptions['defaultRankingTheme'] = {
    primary: PRIMARY_COLOR,
    primaryLight: '#FF6969',
    light: '#FFF5E0',
    dark: '#141E46',
    info: '#9ED9EB'
}

export const RANKING_APP_THEME: Theme = createTheme({
    defaultRankingTheme: THEMES_COLORS,
    palette: {
        primary: {
            main: alpha(PRIMARY_COLOR, 0.7),
            light: alpha(PRIMARY_COLOR, 0.5),
            dark: alpha(PRIMARY_COLOR, 0.9),
        }
    }
});
