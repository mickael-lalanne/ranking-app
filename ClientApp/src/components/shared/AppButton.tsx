import React, { useEffect, useState } from 'react';
import { css } from '@emotion/css';
import { useTheme } from '@mui/material/styles';
import { EResponsiveBreakpoints } from '../../utils/css-utils';
import { useWindowSize } from '../../hooks/useWindowsSize';

const AppButton = (
    { text, icon, onClickHandler, color, disabled = false } : {
        text: string;
        icon?: React.JSX.Element;
        onClickHandler: () => void,
        color?: string,
        disabled?: boolean
    }
) => {
    const theme = useTheme();
    const [btnColor, setBtnColor] = useState<string>(theme.defaultRankingTheme.primary);
    const windowSize = useWindowSize();

    // Change the default button color if a color is provided
    useEffect(() => {
        if (color) {
            setBtnColor(color);
        }
    }, [color]);

    /**
     * @returns {React.JSX.Element | string} the icon if the window size is too smal, otherwise returns the text
     */
    const ButtonContent = (): React.JSX.Element | string => {
        return icon && windowSize.width! < parseFloat(EResponsiveBreakpoints.md)
            ? icon
            : text;
    };

    return(
        <button
            className={`
                ${add_btn_pushable}
                ${disabled ? add_btn_disable : ''}
            `}
            role="button"
            onClick={onClickHandler}
            disabled={disabled}
            data-cy="app-button"
        >
            <span className={add_btn_shadow + ' add_btn_shadow'}></span>
            <span className={app_add_btn_edge} style={{ background: btnColor }}></span>
            <span
                className={`
                    ${app_add_btn_front}
                    add_btn_front
                    ${icon ? add_btn_front_icon : ''}
                `}
                style={{ background: btnColor, color: btnColor === 'white' ? 'black' : 'white' }}
            >
                {ButtonContent()}
            </span>
        </button>
    );
};

export default AppButton;

/**
 * CSS STYLES
 */
const add_btn_pushable = css`
    height: fit-content;
    position: relative;
    border: none;
    background: transparent;
    padding: 0;
    cursor: pointer;
    outline-offset: 4px;
    transition: filter 250ms;
    user-select: none;
    -webkit-user-select: none;
    touch-action: manipulation;
    &:hover {
        filter: brightness(110%);
        -webkit-filter: brightness(110%);
        .add_btn_front {
            transform: translateY(-6px);
            transition:
              transform
              250ms
              cubic-bezier(.3, .7, .4, 1.5);
        }
        .add_btn_shadow {
            transform: translateY(4px);
            transition:
                transform
                250ms
                cubic-bezier(.3, .7, .4, 1.5);
        }
    }
    &:active {
        .add_btn_front {
            transform: translateY(-2px);
            transition: transform 34ms;
        }
        .add_btn_shadow {
            transform: translateY(1px);
            transition: transform 34ms;
        }
    }
    &:focus:not(:focus-visible) {
        outline: none;
    }
`;

const add_btn_shadow = css`
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border-radius: 12px;
    background: hsl(0deg 0% 0% / 0.25);
    will-change: transform;
    transform: translateY(2px);
    transition:
        transform
        600ms
        cubic-bezier(.3, .7, .4, 1);
`;

const app_add_btn_edge = css`
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border-radius: 12px;
    filter: brightness(0.5);
`;

const app_add_btn_front = css`
    display: block;
    position: relative;
    padding: 12px 27px;
    border-radius: 12px;
    font-size: 1.1rem;
    will-change: transform;
    transform: translateY(-4px);
    transition:
        transform
        600ms
        cubic-bezier(.3, .7, .4, 1);
    border: 1px solid black;
`;

const add_btn_front_icon = css({
    padding: '15px 16px'
});

const add_btn_disable = css({
    filter: 'grayscale(1)',
    pointerEvents: 'none',
    userSelect: 'none',
    opacity: 0.5
});
