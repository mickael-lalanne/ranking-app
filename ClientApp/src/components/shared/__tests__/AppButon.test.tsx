import * as React from 'react';
import AppButton, { AppButtonProps } from '../AppButton';
import userEvent from '@testing-library/user-event';
import { renderWithProviders } from '../../../utils/test-utils';
import { fireEvent, screen } from '@testing-library/react';
import CancelIcon from '@mui/icons-material/Cancel';
import { act } from 'react-dom/test-utils';
import { EResponsiveBreakpoints } from '../../../utils/css-utils';

const DEFAULT_PROPS: AppButtonProps = {
    onClickHandler: jest.fn(),
    text: 'My Button text',
};
const AppButtonTest = (props: AppButtonProps) => {
    return (
        <AppButton
            text={props.text}
            onClickHandler={props.onClickHandler}
            disabled={props.disabled}
            icon={props.icon}
        />
    );
};

describe('Screenshot', () => {
    it('Takes default screenshot', () => {
        const component = renderWithProviders(AppButtonTest(DEFAULT_PROPS));

        expect(component.baseElement).toMatchSnapshot();
    });
});

describe('Common', () => {
    it('Clicks on the button', async () => {
        renderWithProviders(AppButtonTest(DEFAULT_PROPS));

        expect(DEFAULT_PROPS.onClickHandler).toHaveBeenCalledTimes(0);
        await userEvent.click(screen.getByText(DEFAULT_PROPS.text));
        expect(DEFAULT_PROPS.onClickHandler).toHaveBeenCalledTimes(1);
    });

    it('checks if disable', () => {
        // By default, the button is enable
        const { rerender } = renderWithProviders(AppButtonTest(DEFAULT_PROPS));
        expect(
            screen.getByText(DEFAULT_PROPS.text).closest('button')
        ).toBeEnabled();

        // Change the disabled props to check if the button is now disabled
        rerender(AppButtonTest({ ...DEFAULT_PROPS, disabled: true }));
        expect(
            screen.getByText(DEFAULT_PROPS.text).closest('button')
        ).toBeDisabled();
    });
});

describe('Responsive', () => {
    it('displays the text', () => {
        renderWithProviders(
            AppButtonTest({ ...DEFAULT_PROPS, icon: <CancelIcon /> })
        );

        expect(screen.getByText(DEFAULT_PROPS.text)).toBeDefined();
    });

    it('displays the icon', async () => {
        renderWithProviders(
            AppButtonTest({ ...DEFAULT_PROPS, icon: <CancelIcon /> })
        );

        // Change the windows size to have a small screen
        await act(async () => {
            Object.defineProperty(window, 'innerWidth', {
                writable: true,
                configurable: true,
                // In jest environment, the size will not be converted in px
                // So we can use directly the em value
                value: parseFloat(EResponsiveBreakpoints.sm),
            });
            await fireEvent(window, new Event('resize'));
        });

        // Now, the text button should not be displayed
        expect(await screen.queryByText(DEFAULT_PROPS.text)).toBeNull();
        // The icon should be displayed instead
        expect(await screen.findByTestId('app-btn-icon')).toBeDefined();
    });
});
