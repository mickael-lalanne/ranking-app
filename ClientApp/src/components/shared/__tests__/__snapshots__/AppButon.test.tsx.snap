// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes default screenshot 1`] = `
<body>
  <div>
    <button
      class="
                css-1t5c0s4
                
            "
      data-cy="app-button"
      role="button"
    >
      <span
        class="css-1uxuv1w add_btn_shadow"
      />
      <span
        class="css-e5grz5"
      />
      <span
        class="
                    css-1pqqf9o
                    add_btn_front
                    
                "
        style="color: white;"
      >
        My Button text
      </span>
    </button>
  </div>
</body>
`;
